src/contracts/logic/math.rs

```diff
@@ -0,0 +1,444 @@
+use decimal::*;
+
+use crate::math::liquidity::Liquidity;
+use crate::math::sqrt_price::sqrt_price::{calculate_sqrt_price, SqrtPrice};
+use crate::math::token_amount::TokenAmount;
+use crate::math::MAX_TICK;
+
+const PRICE_DENOMINATOR: u128 = 1_000_000_000_000_000_000_000_000;
+const LIQUIDITY_DENOMINATOR: u128 = 1_000_000;
+
+// taken from protocol
+const TICK_LIMIT: i32 = 44_364;
+
+pub fn get_liquidity(
+    x: TokenAmount,
+    y: TokenAmount,
+    mut lower_tick: i32,
+    mut upper_tick: i32,
+    current_sqrt_price: SqrtPrice,
+    rounding_up: bool,
+    tick_spacing: u16,
+) -> (TokenAmount, TokenAmount, Liquidity) {
+    if lower_tick < -MAX_TICK {
+        lower_tick = get_min_tick(tick_spacing);
+    }
+    if upper_tick > MAX_TICK {
+        upper_tick = get_max_tick(tick_spacing);
+    }
+
+    let lower_sqrt_price = calculate_sqrt_price(lower_tick).unwrap();
+    let upper_sqrt_price = calculate_sqrt_price(upper_tick).unwrap();
+
+    if upper_sqrt_price < current_sqrt_price {
+        // single token y
+        let (liquidity, estimated_x) = get_liquidity_by_y_sqrt_price(
+            y,
+            lower_sqrt_price,
+            upper_sqrt_price,
+            current_sqrt_price,
+            rounding_up,
+        );
+        return (estimated_x, y, liquidity);
+    } else if current_sqrt_price < lower_sqrt_price {
+        // single token x
+        let (liquidity, estimated_y) = get_liquidity_by_x_sqrt_price(
+            x,
+            lower_sqrt_price,
+            upper_sqrt_price,
+            current_sqrt_price,
+            rounding_up,
+        );
+        return (x, estimated_y, liquidity);
+    }
+    let (liquidity_by_y, estimated_x) = get_liquidity_by_y_sqrt_price(
+        y,
+        lower_sqrt_price,
+        upper_sqrt_price,
+        current_sqrt_price,
+        rounding_up,
+    );
+    let (liquidity_by_x, estimated_y) = get_liquidity_by_x_sqrt_price(
+        x,
+        lower_sqrt_price,
+        upper_sqrt_price,
+        current_sqrt_price,
+        rounding_up,
+    );
+    if liquidity_by_y < liquidity_by_x {
+        (estimated_x, estimated_y, liquidity_by_y)
+    } else {
+        (estimated_x, estimated_y, liquidity_by_x)
```

Comment:
return struct instead of tuple. It's easy to make a mistake and take the wrong token